### PR TITLE
DPO padding-free

### LINKED ISSUE #2469: Packing in DPOTrainer
### Feature request

packing can be supported in dpo trainer.

### Motivation

sequences with high variation in length padded together will waste a lot of resources. issue #1274 mentioned this, but i don't think the conclusion is correct. packing won't conflict with pairwise datasets. you just need to unpack the sequence after forwarding. you can easily identify the start and the end of each sequence by the position of 0 in `position_ids`.

### Your contribution

actually i already have a version of dpo trainer that can deal with packing:

```python
    def _forward_one(self, model: nn.Module, batch: dict, name: str):
        batch = dict(batch)
        targets = batch.pop("targets")
        loss_masks = batch.pop("loss_masks").to(torch.bool)
        out = model(**batch)
        logits = out["logits"]
        targets = targets.clone()
        targets[~loss_masks] = 0  # dummy token; we'll ignore the losses on these tokens later
        per_token_logps = torch.gather(logits.log_softmax(-1), dim=2, index=targets.unsqueeze(2)).squeeze(2)
        per_token_logps[~loss_masks] = 0
        # unpack logp
        assert batch["position_ids"].shape[0] == 1
        starts = (batch["position_ids"] == 0).nonzero()[:, 1]
        seqs = []
        for i in range(len(starts) - 1):
            seqs.append(per_token_logps[0, starts[i]: starts[i + 1]].sum())
        seqs.append(per_token_logps[0, starts[-1]:].sum())
        all_logps = torch.stack(seqs)
        return {
            f"{name}_logps": all_logps,
            f"mean_{name}_logits": logits[loss_masks].mean(),
        }

    def concatenated_forward(self, model: nn.Module, batch: dict):
        ret = {}
        for name in ("chosen", "rejected"):
            ret = {**ret, **self._forward_one(model, batch[name], name)}
        return ret
```

this does not directly fit into the complex logic of dpo trainer, but the idea is that it is possible to do packing.

### PR DESCRIPTION
**Pull request by qgallouedec**
# What does this PR do?

Fixes https://github.com/huggingface/trl/issues/2469

## Before submitting
- [ ] This PR fixes a typo or improves the docs (you can dismiss the other checks if that's the case).
- [ ] Did you read the [contributor guideline](https://github.com/huggingface/trl/blob/main/CONTRIBUTING.md#create-a-pull-request),
      Pull Request section?
- [ ] Was this discussed/approved via a GitHub issue? Please add a link
      to it if that's the case.
- [ ] Did you make sure to update the documentation with your changes?
- [ ] Did you write any new necessary tests?

## Who can review?

Anyone in the community is free to review the PR once the tests have passed. Feel free to tag
members/contributors who may be interested in your PR.

---

<a href="https://app.devin.ai/review/huggingface/trl/pull/5141" target="_blank">
  <picture>
    <source media="(prefers-color-scheme: dark)" srcset="https://static.devin.ai/assets/gh-open-in-devin-review-dark.svg?v=1">
    Open with Devin
  </picture>
</a>

### CODE REVIEW

File: trl/trainer/dpo_trainer.py [hunk 1]
```diff
@@ -1110,13 +1194,22 @@ def _compute_loss(self, model, inputs, return_outputs):
...
+                else:
+                    model_kwargs[key] = inputs[key]
 
         outputs = model(**model_kwargs)
         shift_logits = outputs.logits[..., :-1, :].contiguous()
+        if self.padding_free:
+            shift_logits = restore_padding_from_flattened(shift_logits, flat_position_ids)
         shift_labels = input_ids[..., 1:].contiguous()
         shift_completion_mask = completion_mask[..., 1:].contiguous()
         per_token_logps = selective_log_softmax(shift_logits, shift_labels)
```
Comment (qgallouedec) [inline trl/trainer/dpo_trainer.py:1464]: 🔴 **Dimension mismatch between restored logits and shift_labels when `padding_free=True` with `pad_to_multiple_of`**

When both `padding_free=True` and `pad_to_multiple_of` is set, `restore_padding_from_flattened` produces a tensor whose sequence dimension is `max_unpadded_seq_len - 1`, while `shift_labels` and `shift_completion_mask` (derived from the original padded `input_ids`) have sequence dimension `padded_seq_len - 1`. Because `pad_to_multiple_of` can make `padded_seq_len > max_unpadded_seq_len`, these dimensions will not match, causing a runtime crash in `selective_log_softmax` (or the liger loss).

<details>
<summary>Detailed explanation with concrete example</summary>

Consider a batch with 2 sequences of lengths 3 and 2, and `pad_to_multiple_of=4`:
- DataCollator pads to length 4 → `input_ids` shape: `(2, 4)`
- `shift_labels = input_ids[..., 1:]` → shape `(2, 3)`
- Flattening removes padding → 5 non-padding tokens
- Model forward → `(1, 5, vocab)`, after `[..., :-1, :]` → `(1, 4, vocab)`
- `restore_padding_from_flattened` removes cross-boundary elements and pads back → shape `(2, 2, vocab)` (max unpadded length 3 minus 1 = 2)
- `selective_log_softmax(shift_logits, shift_labels)` receives shapes `(2, 2, vocab)` and `(2, 3)` → **crash**

This affects three methods: `_compute_loss` at `trl/trainer/dpo_trainer.py:1215`, `compute_ref_log_probs` at `trl/trainer/dpo_trainer.py:1080`, and `_compute_loss_liger` at `trl/trainer/dpo_trainer.py:1146-1147`. There is no validation preventing the user from setting both `padding_free=True` and `pad_to_multiple_of`.

**Impact:** Any user combining `padding_free=True` with a non-None `pad_to_multiple_of` will hit a dimension mismatch crash during training whenever `pad_to_multiple_of` rounds up beyond the longest sequence in the batch.

</details>

<details>
<summary>Prompt for agents</summary>

```
In trl/trainer/dpo_trainer.py, around where the padding_free and vision dataset checks are done (near line 635-638), add a validation that raises an error when both padding_free=True and pad_to_multiple_of is not None, since restore_padding_from_flattened reconstructs tensors based on actual (unpadded) sequence lengths, which will be shorter than the padded_seq_len when pad_to_multiple_of rounds up. For example, add after line 638:

if self.padding_free and args.pad_to_multiple_of is not None:
    raise ValueError(
        "Padding-free training is not compatible with `pad_to_multiple_of`. "
        "Please set `pad_to_multiple_of=None` when using `padding_free=True`."
    )

Alternatively, the fix could be applied inside the loss computation methods (_compute_loss, compute_ref_log_probs, _compute_loss_liger) by truncating shift_labels and shift_completion_mask to match the restored logits' sequence length after restore_padding_from_flattened is called.
```

</details>


<a href="https://app.devin.ai/review/huggingface/trl/pull/5141" target="_blank">
  <picture>
    <source media="(prefers-color-scheme: dark)" srcset="https://static.devin.ai/assets/gh-open-in-devin-review-dark.svg?v=1">
    Open in Devin Review
  </picture>
</a>


---
*Was this helpful? React with 👍 or 👎 to provide feedback.*

File: trl/trainer/dpo_trainer.py [hunk 2]
```diff
@@ -90,6 +90,63 @@ def get_dataset_column_names(dataset: Dataset | IterableDataset) -> list[str]:
...
+    Returns:
+        `torch.Tensor`:
+            Restored tensor with shape `(batch_size, max(sequence_lengths - 1), ...)`.
+    """
+    keep_mask = flat_position_ids[:, 1:].ne(0).squeeze(0)
+    tensor = tensor.squeeze(0)[keep_mask]
+    starts = flat_position_ids.squeeze(0).eq(0).nonzero(as_tuple=True)[0]
+    ends = torch.cat((starts[1:], starts.new_tensor([flat_position_ids.size(1)])))
+    split_lengths = (ends - starts - 1).clamp_min(0).tolist()
+    return pad(list(tensor.split(split_lengths, dim=0)), padding_value=padding_value)
```
Comment (qgallouedec) [inline trl/trainer/dpo_trainer.py:145]: 📝 **Info: restore_padding_from_flattened cross-sequence boundary filtering is correct but subtle**

The `keep_mask` at `trl/trainer/dpo_trainer.py:142` filters elements from the shifted (`[..., :-1, :]`) tensor using `flat_position_ids[:, 1:].ne(0)`. This removes exactly the cross-sequence boundary predictions (where a token from sequence N predicts the first token of sequence N+1). For example, with `flat_position_ids = [[0,1,2,3,0,1,0,1,2]]`, the shifted position IDs are `[1,2,3,0,1,0,1,2]`, and `.ne(0)` correctly identifies indices 3 and 5 as boundaries to discard. The `split_lengths` at line 146 use `(ends - starts - 1).clamp_min(0)` which correctly yields per-sequence shifted lengths (original_length - 1). This logic is correct but non-obvious — a comment explaining why `keep_mask` filters cross-boundary predictions would improve maintainability.


<a href="https://app.devin.ai/review/huggingface/trl/pull/5141" target="_blank">
  <picture>
    <source media="(prefers-color-scheme: dark)" srcset="https://static.devin.ai/assets/gh-open-in-devin-review-dark.svg?v=1">
    Open in Devin Review
  </picture>
</a>


---
*Was this helpful? React with 👍 or 👎 to provide feedback.*

File: trl/trainer/dpo_trainer.py (hunk 1 above)
Comment (qgallouedec) [inline trl/trainer/dpo_trainer.py:1464]: 📝 **Info: Padding value of 0 in restored logits is safe due to downstream masking**

In `_compute_loss` at line 1212, `restore_padding_from_flattened(shift_logits, flat_position_ids)` uses the default `padding_value=0`, meaning restored padding positions get all-zero logits. While this produces incorrect log-probabilities at those positions (approximately `-log(vocab_size)` after log_softmax), the subsequent masking at line 1216 (`per_token_logps[shift_completion_mask == 0] = 0.0`) zeros them out before summation. Similarly in `compute_ref_log_probs` at line 1081. So the choice of padding_value=0 is safe, though it could be confusing to future readers.

*(Refers to lines 1211-1216)*


<a href="https://app.devin.ai/review/huggingface/trl/pull/5141" target="_blank">
  <picture>
    <source media="(prefers-color-scheme: dark)" srcset="https://static.devin.ai/assets/gh-open-in-devin-review-dark.svg?v=1">
    Open in Devin Review
  </picture>
</a>


---
*Was this helpful? React with 👍 or 👎 to provide feedback.*

File: tests/test_dpo_trainer.py [hunk 3]
```diff
@@ -132,6 +136,47 @@ def test_with_pad_to_multiple_of(self):
...
+        restored = restore_padding_from_flattened(flattened, flat_position_ids, padding_value=-100)
+
+        expected = torch.tensor(
+            [
+                [[0.0, 1.0], [2.0, 3.0], [4.0, 5.0]],
+                [[8.0, 9.0], [-100.0, -100.0], [-100.0, -100.0]],
+                [[12.0, 13.0], [14.0, 15.0], [-100.0, -100.0]],
+            ]
+        )
+        torch.testing.assert_close(restored, expected)
```
Comment (qgallouedec) [inline tests/test_dpo_trainer.py]: 📝 **Info: Test for restore_padding_from_flattened only covers multi-token sequences**

The test at `tests/test_dpo_trainer.py:164-177` covers sequences of lengths 4, 2, and 3, but does not test edge cases such as a sequence with exactly 1 token (which would produce 0 shifted tokens and an empty split). While tracing the logic shows `clamp_min(0)` at `trl/trainer/dpo_trainer.py:146` handles this gracefully (producing an empty tensor that gets padded), and such short sequences are unlikely in DPO training, an explicit test would increase confidence. Similarly, there's no test for the case where all sequences have the same length.


<a href="https://app.devin.ai/review/huggingface/trl/pull/5141" target="_blank">
  <picture>
    <source media="(prefers-color-scheme: dark)" srcset="https://static.devin.ai/assets/gh-open-in-devin-review-dark.svg?v=1">
    Open in Devin Review
  </picture>
</a>


---
*Was this helpful? React with 👍 or 👎 to provide feedback.*

File: trl/trainer/dpo_trainer.py [hunk 4]
```diff
@@ -1110,13 +1194,22 @@ def _compute_loss(self, model, inputs, return_outputs):
         completion_mask = inputs["completion_mask"]
         input_ids, attention_mask, completion_mask = self._truncate_inputs(input_ids, attention_mask, completion_mask)
 
-        model_kwargs = {"input_ids": input_ids, "attention_mask": attention_mask, "use_cache": False}
+        if self.padding_free:
+            flat_input_ids, flat_position_ids = flatten_batch_for_padding_free(input_ids, attention_mask)
+            model_kwargs = {"input_ids": flat_input_ids, "position_ids": flat_position_ids, "use_cache": False}
+        else:
+            model_kwargs = {"input_ids": input_ids, "attention_mask": attention_mask, "use_cache": False}
```
Comment (qgallouedec) [inline trl/trainer/dpo_trainer.py:1450]: 📝 **Info: No attention_mask passed to model in padding_free mode relies on model auto-inferring it from position_ids**

When `padding_free=True`, the model_kwargs at `trl/trainer/dpo_trainer.py:1199` include `position_ids` but omit `attention_mask`. This relies on the model's flash attention implementation inferring sequence boundaries from position_ids (detecting resets to 0 as new sequence starts). This is the standard convention for HuggingFace flash attention implementations and is correct for the supported flash attention variants listed in `FLASH_ATTENTION_VARIANTS`. However, if a user bypasses the warning and uses a non-flash attention implementation, the lack of attention_mask would cause incorrect self-attention (all tokens attending to all other tokens across sequence boundaries).


<a href="https://app.devin.ai/review/huggingface/trl/pull/5141" target="_blank">
  <picture>
    <source media="(prefers-color-scheme: dark)" srcset="https://static.devin.ai/assets/gh-open-in-devin-review-dark.svg?v=1">
    Open in Devin Review
  </picture>
</a>


---
*Was this helpful? React with 👍 or 👎 to provide feedback.*

**Review by qgallouedec — COMMENTED** [review]
**Devin Review** found 5 potential issues.

View 1 additional finding in Devin Review.


<a href="https://app.devin.ai/review/huggingface/trl/pull/5141" target="_blank">
  <picture>
    <source media="(prefers-color-scheme: dark)" srcset="https://static.devin.ai/assets/gh-open-in-devin-review-dark.svg?v=1">
    Open in Devin Review
  </picture>
</a>

**Review by albertvillanova — COMMENTED** [review]
It does not compile: unclosed raise ValueError( at dpo_trainer.py:787, introduced by the 2026-06-08 main merge (2d3407d5) and carried silently through all three later merges. One-char fix.

**Comment by qgallouedec** [comment]
Closing this one for now.

thks @albertvillanova, you were right, and it's worse than the paren: the tests don't parse either, and the same merge left `_compute_loss` forwarding before it rebuilds `model_kwargs` and `_compute_loss_liger` reading unassigned names.

Main has moved too far since February for a fixup, so I'll redo padding-free DPO from a fresh branch when I get back to it.

**Comment by BrownianNotion** [comment]
@qgallouedec I'd love to give this a shot if you haven't already started working on it!
